### PR TITLE
Reimplement smartredis

### PR DESCRIPTION
**Pull request by ashao**
Significant changes to MOM_MEKE.F90 made it difficult to direcly
merge the initial implementation of SmartRedis with MOM6. A
refactor of the implementation was also done in consultation with
@adcroft to isolate the SmartRedis implementations of the code
which includes creating two stub modules for the
 - SmartRedis client module with associated methods
 - MEKE-related module for inferring EKE using a neural network
   via SmartRedis

The substantive code is now hosted for users interested in
replicating the work at https://github.com/CrayLabs/MOM6-smartredis

### CODE REVIEW

**Comment by ashao** [comment]
This is a draft PR for now since I need to make sure that the implementation still works. I would appreciate feedback from  @adcroft and @marshallward on the overall design to isolate the portions of the codebase that depend on the external SmartRedis package.

@adcroft, I think this largely follows the strategy that we had outlined.

**Comment by marshallward** [comment]
For some reason, Doxygen objects to the `it's` in your docstring for `MEKE_in_dynamics`.

I don't even want to understand this problem.  But replacing with `it is` will fix the docstring error in the subsequent variables.

File: src/parameterizations/lateral/MOM_MEKE.F90 [hunk 1]
```diff
@@ -569,102 +594,116 @@ subroutine step_forward_MEKE(MEKE, h, SN_u, SN_v, visc, dt, G, GV, US, CS, hu, h
...
+            MEKE%Kh(i,j) = (CS%MEKE_KhCoeff * &
+                       sqrt(2.*max(0.,barotrFac2(i,j)*MEKE%MEKE(i,j))*G%areaT(i,j)) ) * &
+                       min(MEKE%Rd_dx_h(i,j), 1.0)
+          enddo ; enddo
         else
           !$OMP parallel do default(shared)
           do j=js,je ; do i=is,ie
             MEKE%Kh(i,j) = CS%MEKE_KhCoeff * &
-                sqrt(2.*max(0., barotrFac2(i,j)*MEKE%MEKE(i,j))) * LmixScale(i,j)
+                sqrt(2.*max(0., barotrFac2(i,j)*MEKE%MEKE(i,j))*G%areaT(i,j))
```
Comment (adcroft) [inline src/parameterizations/lateral/MOM_MEKE.F90:630]: I'm not sure if GH is reporting this right. Not sure why adding SmartRedis involves replacing LmixScale with sqrt(areaT)?

File: src/parameterizations/lateral/MOM_MEKE.F90 [hunk 2]
```diff
@@ -5,27 +5,42 @@ module MOM_MEKE
...
+use MOM_smartredis_MEKE,  only : smartredis_meke_CS_type, smartredis_meke_init, infer_meke
+use MOM_restart,          only : MOM_restart_CS, register_restart_field, query_initialized
+use MOM_smartredis,       only : client_type, smartredis_CS_type
+use MOM_string_functions, only : lowercase
+use MOM_unit_scaling,     only : unit_scale_type
+use MOM_variables,        only : vertvisc_type, thermo_var_ptrs
+use MOM_verticalGrid,     only : verticalGrid_type
+use MOM_MEKE_types,       only : MEKE_type
+
+use time_interp_external_mod, only : init_external_field, time_interp_external
```
Comment (adcroft) [inline src/parameterizations/lateral/MOM_MEKE.F90:28]: Please use MOM_interpolate rather that time_interp_external_mod. We no longer allow direct calls to the infrastructure from src code, but instead require all infrastructure calls to pass through the modules in src/framework.

File: src/framework/MOM_smartredis.F90 [hunk 3]
```diff
@@ -0,0 +1,78 @@
+!> Contains routines necessary to initialize the SmartRedis client
+module MOM_smartredis
+
```
Comment (adcroft) [inline src/framework/MOM_smartredis.F90:3]: Needs the MOM6 license one-liner

File: config_src/external/smartredis/smartredis_client.F90 [hunk 4]
```diff
@@ -0,0 +1,925 @@
+module smartredis_client
+
```
Comment (adcroft) [inline config_src/external/smartredis/smartredis_client.F90:2]: Needs the MOM6 license one-liner.  This is the stub which lives in MOM6 so must hae the MOM6 license. The actual code that replaces the stub can use whatever license you like.

File: config_src/external/smartredis/MOM_smartredis_MEKE.F90 [hunk 5]
```diff
@@ -0,0 +1,61 @@
+!> Contains routines that contain dummy routines for the smart
+module MOM_smartredis_meke
+
```
Comment (adcroft) [inline config_src/external/smartredis/MOM_smartredis_MEKE.F90:3]: Needs the MOM6 license one-liner.  This is the stub which lives in MOM6 so must hae the MOM6 license. The actual code that replaces the stub can use whatever license you like.

File: src/parameterizations/lateral/MOM_MEKE.F90 (hunk 1 above)
Comment (adcroft) [reply src/parameterizations/lateral/MOM_MEKE.F90:630]: Indeed, it looks like the whitespce changes are confusing GH here.

**Review by adcroft — CHANGES_REQUESTED** [review]
Structurally good. Some minor changes requested

File: src/core/MOM.F90 [hunk 6]
```diff
@@ -1798,7 +1811,7 @@ subroutine initialize_MOM(Time, Time_init, param_file, dirs, CS, restart_CSp, &
   ! This include declares and sets the variable "version".
 # include "version_variable.h"
 
-  integer :: i, j, is, ie, js, je, isd, ied, jsd, jed, nz
+  integer :: i, j, k, is, ie, js, je, isd, ied, jsd, jed, nz
```
Comment (adcroft) [inline src/core/MOM.F90:1814]: Is this `k` used? I don't see new code that uses it...

File: config_src/external/database_comms/README.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+# File description
+
+- `MOM_smartredis.F90` contains just method signatures and elements of the
+  control structure that are imported elsewhere within the primary MOM6
+  code. This includes: `dbclient_CS_type`, `dbclient_type`, and `dbclient_init`
+
+- `smartredis_client.F90` is a skeleton of the actual SmartRedis library
+  used to ensure that the interfaces to the library are maintained without
+  requiring MOM6 users to compile in the the full library
```
Comment (marshallward) [inline config_src/external/database_comms/README.md:23]: Do the filenames in this README need to be updated?

**Review by marshallward — APPROVED** [review]
The SmartRedis communication layer has been well-abstracted and there is a strong logical separation between the the DB and the MEKE solver code.  Overall I think this is a very good design and a good template for future external integration.

There are some minor issues related to the use of `put_tensor()` which accepts C types (specifically `c_float`) but the current equivalent with `real32` is reasonably safe and we can address this in the future if the need arises.

File: config_src/external/database_comms/README.md (hunk 7 above)
Comment (ashao) [reply config_src/external/database_comms/README.md:23]: Yes thank you. This has been fixed.

Comment (marshallward) [reply config_src/external/database_comms/README.md:23]: Actually looks like the second file is still misnamed, and the first file is missing `.F90`.

I would do it myself, but you also need to update the docstrings for `put_tensor` and `unpack_tensor` functions anyway, so might as well do them together.

**Comment by marshallward** [comment]
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/16294 :x: 

This is not passing our regression, due to assumed-rank in `database_client_interface.F90`, which are not currently supported by PGI.

Since MOM6 is nominally a Fortran 2003/2008 program, these will need to be replaced with something older and clunkier.

**Review by adcroft — APPROVED** [review]

**Comment by marshallward** [comment]
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/16346 :heavy_check_mark: :yellow_circle:

**Pull request merged**